crates/shared/src/price_estimation/competition.rs

```diff
@@ -182,24 +185,37 @@ fn successes<R, E>(results: &[(EstimatorIndex, Result<R, E>)]) -> usize {
     results.iter().filter(|(_, result)| result.is_ok()).count()
 }
 
-impl PriceEstimating for RacingCompetitionEstimator<Arc<dyn PriceEstimating>> {
+impl PriceEstimating
+    for RacingCompetitionEstimator<Arc<dyn PriceEstimating>, Arc<dyn NativePriceEstimating>>
+{
     fn estimate(&self, query: Arc<Query>) -> futures::future::BoxFuture<'_, PriceEstimateResult> {
-        self.estimate_generic(
-            query.clone(),
-            query.kind,
-            |estimator, query| estimator.estimate(query),
-            move |a, b| {
-                if is_second_quote_result_preferred(query.as_ref(), a, b) {
-                    Ordering::Less
-                } else {
-                    Ordering::Greater
-                }
-            },
-        )
+        async {
+            let out_token = match query.kind {
+                OrderKind::Buy => query.sell_token,
+                OrderKind::Sell => query.buy_token,
+            };
+            let native_price = self.native.estimate_native_price(out_token).await?;
```

Comment:
Can we move this await into the `compare_results` lambda and delay it as much as possible? This should increase our chances of not making estimates slower (the actual quote can run in parallel with the native estimate).

Reply:
Tried to contain the damage (i.e. additional complexity) the best I could.
Instead of calling the native price estimator and gas estimator inside the generic function (which didn't work nicely) I opted for another argument in the generic function that takes a future that provides the needed context.
Now the supplied comparison function takes 2 results and this context.
The context is returned from a fused future (so we can await it multiple times) that awaits the result of a task that was spawned in the background so it does not delay the actual price estimation requests.